test/test.ml

```diff
@@ -11,205 +11,128 @@
-  purpose with or without fee is hereby granted, provided that the above
-  copyright notice and this permission notice appear in all copies.
-
-  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
-  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
-  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
-  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
-  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
-  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
-  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
----------------------------------------------------------------------------*)
-
-open Kcas;;
-open Printf;;
-
-let nb_iter = 10000;;
-let wait_time = 2;;
-let th1_success = Stdlib.ref true;;
-let th2_success = Stdlib.ref true;;
-let th3_success = Stdlib.ref true;;
-let th4_success = Stdlib.ref true;;
-let th5_success = Stdlib.ref true;;
-
-let v_x = 0;;
-let v_y = 1;;
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+  ---------------------------------------------------------------------------*)
+
+let nb_iter = 100_000
+
+let test_set () =
+  let a = Kcas.ref 0 in
+  let v1 = Kcas.get a in
+  assert (v1 == 0);
+  Kcas.set a 1;
+  let v2 = Kcas.get a in
+  assert (v2 == 1)
 
 let thread1 (a1, a2) =
-  let c1 = [mk_cas a1 v_x v_y ; mk_cas a2 v_x v_y] in
-  let c2 = [mk_cas a1 v_y v_x ; mk_cas a2 v_y v_x] in
+  let c1 = [ Kcas.mk_cas a1 0 1; Kcas.mk_cas a2 0 1 ] in
+  let c2 = [ Kcas.mk_cas a1 1 0; Kcas.mk_cas a2 1 0 ] in
   for _ = 1 to nb_iter do
-    let out1 = kCAS c1 in
-    let out2 = kCAS c2 in
-    if out1 <> true || out2 <> true then begin
-(*      print_endline (sprintf "TH%d  APPEL N°%d ECHEC OUT1 = %b    OUT2 = %b!!!!!!!!!" (Domain.self ()) i out1 out2);*)
-      th1_success := false
-    end
+    let out1 = Kcas.kCAS c1 in
+    let out2 = Kcas.kCAS c2 in
+    assert out1;
+    assert out2
   done;
-  ignore @@ kCAS c1;
-  ()
-;;
+  ignore @@ Kcas.kCAS c1
 
 let thread2 (a1, a2) =
-  let c1 = [mk_cas a1 v_y v_x ; mk_cas a2 v_x v_y] in
-  let c2 = [mk_cas a1 v_x v_y ; mk_cas a2 v_y v_x] in
+  let c1 = [ Kcas.mk_cas a1 1 0; Kcas.mk_cas a2 0 1 ] in
```

Comment:
It seems that if the order of cas operations in the list is reversed
```diff
-  let c1 = [ Kcas.mk_cas a1 1 0; Kcas.mk_cas a2 0 1 ] in
+  let c1 = [ Kcas.mk_cas a2 0 1; Kcas.mk_cas a1 1 0 ] in
```
the test gets stuck.  I could imagine a library like kcas where that would be by design (IOW, the user is required to order the cas operations consistently).  Is this is a bug or not?

Reply:
Not a bug as long as we consider this library a faithful implementation of the kcas paper, which expects all casns to follow the same total order.  
>The update addresses lie in the data section of memory and are held according to some total order agreed by all threads to guarantee non-blocking behaviour,

But yeah, I did not realize, even though it seems logical in hindsight. I'll add a check for this.